server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -941,8 +944,20 @@ public Collection<Reference> insertColumns(List<Reference> columns) {
         }
         List<Reference> newColumns = new ArrayList<>(columns);
         for (var synthetic : undeterministic) {
-            if (synthetic.ref.column().isRoot() && !newColumns.contains(synthetic.ref)) {
-                newColumns.add(synthetic.ref);
+            if (synthetic.ref.column().isRoot()) {
+                if (newColumns.contains(synthetic.ref) == false) {
+                    newColumns.add(synthetic.ref);
+                }
+            } else {
+                int rootIndex = Reference.indexOf(newColumns, synthetic.ref.column().getRoot());
```

Comment:
minor: you could extract `synthetic.ref.column()` or `synthetic.ref.column().getRoot()` to a var.